arangod/Aql/Projections.cpp

```diff
@@ -101,6 +107,175 @@ uint16_t Projections::coveringIndexPosition(
                                  "unable to determine covering index position");
 }
 
+void Projections::produceFromDocument(
+    velocypack::Builder& b, velocypack::Slice slice,
+    transaction::Methods const* trxPtr,
+    fu2::unique_function<void(Variable const*, velocypack::Slice) const> const&
+        cb) const {
+  TRI_ASSERT(slice.isObject());
+
+  size_t levelsOpen = 0;
+
+  for (auto const& it : _projections) {
+    TRI_ASSERT(it.variable != nullptr);
+
+    if (it.type == AttributeNamePath::Type::IdAttribute) {
+      // projection for "_id"
+      TRI_ASSERT(levelsOpen == 0);
+      TRI_ASSERT(it.path.size() == 1);
+
+      b.clear();
+      b.add(VPackValue(transaction::helpers::extractIdString(trxPtr->resolver(),
+                                                             slice, slice)));
+      cb(it.variable, b.slice());
+    } else if (it.type == AttributeNamePath::Type::KeyAttribute) {
+      // projection for "_key"
+      TRI_ASSERT(levelsOpen == 0);
+      TRI_ASSERT(it.path.size() == 1);
+      b.clear();
```

Comment:
Not sure why we have to clear `b` here.

Reply:
`b` is just a temporary scratch buffer here.
The function only uses `b` to write the `_id` value into the Builder, and then calls the callback with a VPackSlice pointing to that Builder.
Because at most one `_id` value is in the Builder, we should clear it before writing into it.
The caller does not expect anything to be in `b` when the function returns, so we should be ok.